modules/generic/generic.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package generic
+
+import (
+	"sync"
+
+	"github.com/opentracing/opentracing-go"
+
+	"go.uber.org/fx/modules"
+	"go.uber.org/fx/service"
+	"go.uber.org/fx/ulog"
+)
+
+// PopulateStruct populates the given configuration struct using the given Controller.
+func PopulateStruct(controller Controller, config interface{}) error {
+	return controller.Host().Config().Scope("modules").Get(controller.Name()).PopulateStruct(config)
+}
+
+// Controller holds data for a Module, and provides functionality
+// for the Module to notify the wrapping service.Module of state changes.
+// TODO(pedge): not the best name
+// TODO(pedge): split up into read-only and read-write interfaces?
+type Controller interface {
+	Host() service.Host
+	Roles() []string
+	Name() string
+	Tracer() opentracing.Tracer
+	Log() ulog.Log
+	// NotifyStopped will result in the wrapping service.Module's
+	// IsRunning() method to return false.
+	NotifyStopped()
+}
+
+// Module is a simpliciation of the service.Module interface
+// that can be wrapped with a service.Module for easier implemenation.
+type Module interface {
```

Comment:
That's fine per golang idioms - the initial intent would be to name it generic.GenericModule, but this fails golang linters because you shouldnt have foo.FooWhatever.